src/app/AiPlayer.ts

```diff
@@ -0,0 +1,195 @@
+        [
+          [AnimalNames.PIG, 1],
+          [AnimalNames.SHEEP, 3],
+          [AnimalNames.RABBIT, 6],
+        ],
+      ],
+      [
+        AnimalNames.PIG,
+        [
+          [AnimalNames.SHEEP, 1],
+          [AnimalNames.RABBIT, 6],
+        ],
+      ],
+      [AnimalNames.SHEEP, [[AnimalNames.RABBIT, 6]]],
+    ];
+    for (const [animalToSell, target] of oneToManyTrades) {
+      if (this.herd.getAnimalNumber(animalToSell) > 1) {
+        const offer: Offer[] = [[animalToSell, 1]];
+        if (
+          gameController.theGame.theTrade.processOffer(
+            offer,
+            this,
+            target,
+          )
+        ) {
+          return;
+        }
+      } else if (this.herd.getAnimalNumber(animalToSell) === 0) {
+        break;
+      }
+    }
+
+    //Try to win by trading many to one (possible if only one kind of animal is missing)
+    const animalsInOrder = [
+      AnimalNames.HORSE,
+      AnimalNames.COW,
+      AnimalNames.PIG,
+      AnimalNames.SHEEP,
+      AnimalNames.RABBIT,
+    ];
+
+    const missingAnimals = animalsInOrder.filter(
+      (animal) => this.herd.getAnimalNumber(animal) === 0,
+    );
+
+    if (missingAnimals.length === 1) {
+      const offer = this.animalsForTrade();
+      const target: Offer[] = [[missingAnimals[0], 1]];
+      if (
+        gameController.theGame.theTrade.processOffer(
+          offer,
+          this,
+          target,
+        )
+      ) {
+        return;
+      }
+    }
+    //Cannot win game by trade. Check if buying dogs is worth it
+    if (this.buyABigDog()) {
+      if (
+        gameController.theGame.theTrade.processOffer(
+          [[AnimalNames.PIG, 3]],
+          this,
+          [[AnimalNames.BIG_DOG, 1]],
+        )
+      ) {
+        return;
+      } else if (
+        gameController.theGame.theTrade.processOffer(
+          [[AnimalNames.COW, 1]],
+          this,
+          [[AnimalNames.BIG_DOG, 1]],
+        )
+      ) {
+        return;
+      }
+    }
+
+    if (this.buyASmallDog()) {
+      const offer: Offer[] = [[AnimalNames.RABBIT, 6]];
+      const target: Offer[] = [[AnimalNames.SMALL_DOG, 1]];
+      if (
+        gameController.theGame.theTrade.processOffer(
+          offer,
+          this,
+          target,
+        )
+      ) {
+        return;
+      }
+    }
+
+    //If we don't need any dogs, try to buy the most expensive animal we can afford
+    for (const missingAnimal of missingAnimals) {
+      const offer = this.animalsForTrade(missingAnimal);
+      const target: Offer[] = [[missingAnimal, 1]];
+      if (
+        gameController.theGame.theTrade.processOffer(
+          offer,
+          this,
+          target,
+        )
+      ) {
+        return;
+      }
+    }
+  }
+
+  private animalsForTrade(
+    targetAnimal: AnimalNames = AnimalNames.HORSE,
+  ): [AnimalNames, number][] {
+    const animalsForTrade: [AnimalNames, number][] = [];
+    this.herd.theAnimals.forEach(([animal, count]) => {
```

Comment:
This method could be a great opportunity to use array.reduce and return the total value from accumulator. Consider a refactor.